module/zfs/dsl_scan.c

```diff
@@ -3044,6 +3045,8 @@ scan_io_queues_run_one(void *arg)
 
 	list_create(&sio_list, sizeof (scan_io_t),
 	    offsetof(scan_io_t, sio_nodes.sio_list_node));
+	queue->q_zio = zio_null(queue->q_scn->scn_zio_root,
+	    queue->q_scn->scn_dp->dp_spa, NULL, NULL, NULL, ZIO_FLAG_CANFAIL);
```

Comment:
Even though it shouldn't be needed it seems prudent to move all the `queue->q_zio` accesses under the `&queue->q_vd->vdev_scan_io_queue_lock`.  As I understand it this shouldn't add much/any contention since the code already takes the lock, and it was the zio lock not the scan_io_queue lock which was contended anyway.  This would also keep the code in line with the `vdev` struct comment.

```c
        /*
         * Protects the vdev_scan_io_queue field itself as well as the
         * structure's contents (when present).
         */
        kmutex_t                        vdev_scan_io_queue_lock;
        struct dsl_scan_io_queue        *vdev_scan_io_queue;
```

If you're really worried about it you could keep the `zio_null()` and `zio_nowait()` calls outside the lock and just move the new  `queue->q_zio` loads/stores under the lock,

Reply:
queue->q_zio field is accessed only from the single thread thread that first assigns it, then uses it to issue the I/Os, and then wipes it.  With the same effect I could pass it through the arguments. There is nothing really to protect with the lock. I can do it if you insist, it should not be an issue, but it should not buy us anything, at least I don't see what.

Reply:
Oh I entirely agree, the lock shouldn't be needed at all the way it's currently used.  My point was just that moving the new code under the lock 1) shouldn't cost us anything, 2) is a more defensive coding style, and 3) aligns the code with the comment.  More importantly, and it might just be me, but when casually reading the code this jumps out at me as suspicious and potentially wrong since it's part of a shared structure yet being accessed outside the lock.  That said, I don't feel strongly about it since it is correct.  I just wanted to mention it because I did give me pause when reviewing the PR to convince myself it was safe.

Reply:
I am not convinced that pretending we are using lock for something while we are not is good.  But whatever...  Done.